File: cpp/cmd/echo.cpp
// SPDX-License-Identifier: MIT
#include <koru/braam.hpp>

Task<i32> proc_main(Args args)
{
    // Full rather than Auto: the whole output is one flush, so the buffering
    // question is not worth a tty_of to answer.
    File::stdout().set_buffering(Buffering::Full);

    usize i      = 1;
    bool newline = true;
    if (i < args.size() && args[i] == "-n") {
        newline = false;
        i++;
    }

    for (bool first = true; i < args.size(); i++, first = false) {
        if (!first && (co_await File::stdout().write(" ")).is_err())
            co_return 1;
        if ((co_await File::stdout().write(args[i])).is_err())
            co_return 1;
    }
    if (newline && (co_await File::stdout().write("\n")).is_err())
        co_return 1;

    co_return (co_await File::stdout().flush()).is_err() ? 1 : 0;
}

File: rust/runtime/examples/echo.rs
// SPDX-License-Identifier: MIT

//! Braam's `src/cmd/echo.cpp`, in Rust — the same program against the same
//! function names, written the way Rust would write it. `cpp/cmd/echo.cpp` is
//! Braam's own source compiled against koru, and scripts/portability.sh runs
//! the two and compares the bytes.

use koru::{Args, Buffering, File};

#[koru::main]
async fn main(args: Args) -> i32 {
    // Full rather than Auto: the whole output is one flush.
    File::stdout().set_buffering(Buffering::Full);

    let mut i = 1;
    let newline = !(i < args.size() && &args[i] == "-n");
    if !newline {
        i += 1;
    }

    for (n, w) in (i..args.size()).enumerate() {
        if n > 0 && File::stdout().write(" ").await.is_err() {
            return 1;
        }
        if File::stdout().write(&args[w]).await.is_err() {
            return 1;
        }
    }
    if newline && File::stdout().write("\n").await.is_err() {
        return 1;
    }

    i32::from(File::stdout().flush().await.is_err())
}
